src/db/gc/mod.rs

```diff
@@ -0,0 +1,309 @@
+        // Don't proceed with next steps until we advance at least `depth` epochs. Sleep and yield
+        // to the main loop in order to refresh the heaviest tipset value.
+        if epochs_since_marked < depth {
+            time::sleep(self.block_time * (depth - epochs_since_marked) as u32).await;
+            return anyhow::Ok(());
+        }
+
+        info!("filter keys for GC");
+        self.filter(tipset, depth).await?;
+
+        info!("GC sweep");
+        self.sweep()?;
+
+        // Make sure we don't run the GC too often.
+        time::sleep(interval).await;
+
+        anyhow::Ok(())
+    }
+}
+#[cfg(test)]
+mod test {
+    use crate::blocks::{BlockHeader, Tipset};
+    use crate::chain::ChainStore;
+
+    use crate::db::{GarbageCollectable, MarkAndSweep, MemoryDB};
+    use crate::message_pool::test_provider::{mock_block, mock_block_with_parents};
+    use crate::networks::ChainConfig;
+
+    use crate::utils::db::CborStoreExt;
+
+    use core::time::Duration;
+
+    use crate::shim::clock::ChainEpoch;
+    use fvm_ipld_blockstore::Blockstore;
+    use std::sync::Arc;
+
+    fn insert_unrechable(db: impl Blockstore, quantity: u64) {
+        for idx in 0..quantity {
+            let block: BlockHeader = mock_block(1 + idx, 1 + quantity);
+            db.put_cbor_default(&block).unwrap();
+        }
+    }
+
+    fn run_to_epoch(db: impl Blockstore, cs: Arc<ChainStore<MemoryDB>>, epoch: ChainEpoch) {
+        let mut heaviest_tipset = cs.heaviest_tipset();
+
+        for _ in heaviest_tipset.epoch()..epoch {
+            let block2 = mock_block_with_parents(heaviest_tipset.as_ref(), 1, 1);
+            db.put_cbor_default(&block2).unwrap();
+
+            let tipset = Arc::new(Tipset::from(&block2));
+            cs.set_heaviest_tipset(tipset).unwrap();
+            heaviest_tipset = cs.heaviest_tipset();
+        }
+    }
+
+    #[tokio::test]
+    async fn test_populate() {
+        let interval = Duration::from_secs(0);
+        let db = Arc::new(MemoryDB::default());
+        let chain_config = Arc::new(ChainConfig::default());
+        let gen_block: BlockHeader = mock_block(1, 1);
+        let depth = 1;
+        db.put_cbor_default(&gen_block).unwrap();
+        let cs = Arc::new(
+            ChainStore::new(db.clone(), db.clone(), chain_config, gen_block.clone()).unwrap(),
+        );
+
+        let cs_cloned = cs.clone();
+        let get_heaviest_tipset = Box::new(move || cs_cloned.heaviest_tipset());
+        let mut gc = MarkAndSweep::new(db.clone(), get_heaviest_tipset, depth, interval);
+
+        // test insufficient epochs
+        gc.gc_workflow(interval).await.unwrap();
+        assert!(gc.marked.is_empty());
+
+        // test marked
+        run_to_epoch(db, cs, depth);
+        gc.gc_workflow(interval).await.unwrap();
+        assert_eq!(gc.marked.len(), 2);
+        assert_eq!(gc.epoch_marked, 1);
+    }
+
+    #[tokio::test]
+    async fn test_filter_and_sweep() {
```

Comment:
This is a rather long test. Can it be broken down into smaller tests?

Reply:
It's a long test, because we need to `mark` before `filter` and `sweep`. It can be refactored into smaller things, but it's still going to have to run them all.

Reply:
It's hard for me to see exactly what you're testing here. If this test started failing, I wouldn't know what was wrong. If it can't be broken down into smaller pieces, could you describe exactly what you're testing and why you believe the assertions must hold true?